File: 53. Maximum Subarray.cpp
// Time: O(n), space: O(1)
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxs = INT_MIN;
        int s = 0;
        for (int num : nums) {
            s += num;
            maxs = max(s, maxs);
            s = max(s, 0);
        }
        
        return maxs;
    }
};

File: 53. Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxsum, subsum = float('-inf'), 0
        for num in nums:
            subsum += num
            # updating maxsum must follow the addition to subsum
            # it cannot be placed after the reset of subsum to 0
            # otherwise it'll fail for case like nums = [-1]
            maxsum = max(maxsum, subsum)
            subsum = max(0, subsum)
        return maxsum
